server/src/domain/asset/asset.service.ts

```diff
@@ -414,32 +416,71 @@ export class AssetService {
     const { ids, removeParent, dateTimeOriginal, latitude, longitude, ...options } = dto;
     await this.access.requirePermission(auth, Permission.ASSET_UPDATE, ids);
 
+    const stackIdsToCheckForDelete: string[] = [];
     if (removeParent) {
-      (options as Partial<AssetEntity>).stackParentId = null;
+      (options as Partial<AssetEntity>).stack = null;
       const assets = await this.assetRepository.getByIds(ids);
+      stackIdsToCheckForDelete.push(...new Set(assets.filter((a) => !!a.stackId).map((a) => a.stackId!)));
       // This updates the updatedAt column of the parents to indicate that one of its children is removed
       // All the unique parent's -> parent is set to null
-      ids.push(...new Set(assets.filter((a) => !!a.stackParentId).map((a) => a.stackParentId!)));
+      await this.assetRepository.updateAll(
+        assets.filter((a) => !!a.stack?.primaryAssetId).map((a) => a.stack!.primaryAssetId!),
+        { updatedAt: new Date() },
+      );
     } else if (options.stackParentId) {
+      //Creating new stack if parent doesn't have one already. If it does, then we add to the existing stack
       await this.access.requirePermission(auth, Permission.ASSET_UPDATE, options.stackParentId);
-      // Merge stacks
-      const assets = await this.assetRepository.getByIds(ids);
-      const assetsWithChildren = assets.filter((a) => a.stack && a.stack.length > 0);
-      ids.push(...assetsWithChildren.flatMap((child) => child.stack!.map((gChild) => gChild.id)));
+      const primaryAsset = await this.assetRepository.getById(options.stackParentId);
+      if (!primaryAsset) {
+        throw new Error('Asset not found');
+      }
+      let stack = primaryAsset.stack;
 
-      // This updates the updatedAt column of the parent to indicate that a new child has been added
-      await this.assetRepository.updateAll([options.stackParentId], { stackParentId: null });
-    }
+      ids.push(options.stackParentId);
+      const assets = await this.assetRepository.getByIds(ids);
+      console.log(
+        'assets',
+        assets.map((a) => a.id),
+      );
+      stackIdsToCheckForDelete.push(
+        ...new Set(assets.filter((a) => !!a.stackId && stack?.id !== a.stackId).map((a) => a.stackId!)),
+      );
+      const assetsWithChildren = assets.filter((a) => a.stack && a.stack.assets.length > 0);
+      ids.push(...assetsWithChildren.flatMap((child) => child.stack!.assets.map((gChild) => gChild.id)));
+
+      if (!stack) {
+        stack = await this.assetStackRepository.create({
+          primaryAssetId: primaryAsset.id,
+          assets: ids.map((id) => {
+            return { id } as AssetEntity;
+          }),
+        });
+      } else {
+        await this.assetStackRepository.save({
+          id: stack.id,
+          primaryAssetId: primaryAsset.id,
+          assets: ids.map((id) => {
+            return { id } as AssetEntity;
+          }),
+        });
```

Comment:
I'm a little thrown off by the logic... We're in the branch where option has a `stackParentId`. `primaryAsset` is exactly this parent. Now we get the stack from `primaryAsset` and check if that is defined. But that should always be the case because otherwise the `primaryAsset` couldn't have been found (using the stack) in the first place, no?

Reply:
On my phone, but iirc we are in the branch where options.parentId is set, which is setting a new parent. If this pare t doesn't already have a stack we must create on, make it the parent, then delete all the other stacks that were merged into this new stack.

Reply:
Yea this was correct. Options is passed to us by the client, options.stackParentId is the new parent to be assigned to all the asset ids passed in the request. So we must create a new stack if one doesn't exist, or merge all other assets into the stack that this asset is currently in. In both cases, we set the primaryAssetId of the stack, new or existing, to the options.stackParentId (but we use primaryAsset.id), these values will always be the same. After this is done we then cleanup any stacks that no longer have assets later in the function.

Reply:
Got it. Thanks for the explanation! :)